### DISCUSSION TITLE
Dual Screen Support for ST7789 SPI

### DISCUSSION THREAD

**Discussion by jlt206** [Bodmer/TFT_eSPI#2576] [General]
Does Dual Screen Support exist for ST7789 SPI?

Thanks!

**Comment by Bodmer** [comment]
It is possible but the display chip selects must be managed by the sketch.  Some time ago I noticed a youtube video describing this.

**Reply by jlt206** [reply]
Hi, Yes. I found the video which is very good here ( https://www.youtube.com/live/cCgNHIHijhs?feature=share ) and example code is here on Git ( https://github.com/kitecraft/Multiple_TFT_Displays ).

So I will use these to meka it work. 

What I did find as one issue is the amount of memory required which, is significant and requires more than just a Arduino Uno/Nano so I have to find another solution. In the video a ESP32 is used, but with the other functions I require as well, I don't think that will be enough.

**Reply by Bodmer** [reply]
The RaSpberry Pico RP2040 board is my favorite. A Teensy 4 can be used too and has a high processor speed with lots of RAM.

**Reply by jlt206** [reply]
Yes. I found the Teensy 4 which looks very good. I will also look at the RaSpberry Pico RP2040 board as you suggest.

Thanks!

**Reply by jlt206** [reply]
So... I have attempted this configuration but so far without success.
I am not able to display on both screens and am having trouble even getting one screen to display with basic dual screen code.
I have adjusted the user setup files as suggested.

Any ideas as I can only think I am missing something else in the conifiguration files?

I am using an ESP8266 NodeMCU and 2 x the same 7789 Pimorino 240x240 screens.

I can make only make each screen work by moving the CS min from each to PIN D8.

I have used;
D3 - DC.    - Orange 
D5 - SCK   - Green
D7 - MOSI - Yellow

D8 - CS1   - Blue
D4 - CS2  - Blue

![PXL_20230416_104002887](https://user-images.githubusercontent.com/4954838/232301421-6afed390-c219-43c4-a83b-dd4f5f24ca7e.jpg)

.h
[User_Setup_Select.txt](https://github.com/Bodmer/TFT_eSPI/files/11241982/User_Setup_Select.txt)

.ino 
[Basic_Double_TFT.txt](https://github.com/Bodmer/TFT_eSPI/files/11241987/Basic_Double_TFT.txt)

**Reply by Bodmer** [reply]
Post Dual_PimoRound.h file

**Comment by jlt206** [comment]
[Dual_PimoRound.txt](https://github.com/Bodmer/TFT_eSPI/files/11243834/Dual_PimoRound.txt)


I have been trying all kinds of variations but still no success. I can only get one screen to work which is always the second and the first occaisionally appears in a ghost format but never using the assigned PIN, except D8-15.

**Reply by Bodmer** [reply]
GPIO and Dx numbers are not the same, in sketch use:
int firstScreenCS = PIN_D4;
int secondScreenCS = PIN_D8;

The mapping is board dependant, for the NodeMCU boards:
https://github.com/Bodmer/TFT_eSPI/blob/master/User_Setup_Select.h#L280

**Reply by jlt206** [reply]
Hi, Thanks for your help.

I made the necessary changes and confirmed that both PIN_D2 and PIN_D8 both work correctly and display screen 1 and 2 correctly fo the pin.

However, I now have run into the problem of only one screen working and it seem to only be screen 2, when I connect both screens.

When I run the code it seems to display screen 1 first then overwrite it with screen 2.

![PXL_20230417_102300047](https://user-images.githubusercontent.com/4954838/232460034-9c6dc573-b27c-4977-826b-cc79dae84927.jpg)

**Comment by Bodmer** [comment]
The screens do not have a hardware reset pin unfortunately. This means if a screen receives garbage commands it can "lock up" and then fail to respond to a software  reset or other commands.   The ESP8266 and ESP32 output clock signals at boot and if these pins are used for chip selects the display can lock up. Try other pins for the chip selects and add pullups so they are not floating at boot time. Check ESP8266 documentation or info websites.

You may find disconnecting power and reconnecting after sketch upload makes both displays work as then each display gets a "power on reset".

So many gotchas!

**Reply by jlt206** [reply]
OK, I'll see what I can find. It does seem that this is a challenging combination.

**Reply by Bodmer** [reply]
I found this link:
https://randomnerdtutorials.com/esp8266-pinout-reference-gpios/

Scroll down to see a nice table of good pins to use.

D8 is pulled low at boot so display could receive garbage. Looks like D2, D3 and D4 are good for chip selects and possibly D0 as that is high at boot time.

**Reply by jlt206** [reply]
Thanks... I've been looking around and I think I may have reached a technical limit with this particular display. So, I may have to select an alternative instead.

**Reply by jlt206** [reply]
Hi, So. I will change to the ESP32 and GCA01 Displays to get it to work fully but for now I will continue to develop on one screen until then.

But for now I am struggling with Sprites. I can't seem to get them working at all.

Declarations like this ----> TFT_eSprite INTERNETimageSprite= TFT_eSprite(&tft);

Setup Like this like this ----> 

void setup()
{
 Serial.begin(115200);
  tft.setSwapBytes(true);
  INTERNETimageSprite.setColorDepth(4);
  INTERNETimageSprite.setSwapBytes(true);
  INTERNETimageSprite.createSprite(160,100);

Push like this ----> INTERNETimageSprite.pushImage(40,40,160,100,(uint16_t *)Internet_image);

Image Converted like this ---->
#include "TFT_eSPI.h"
// 'Internet', 160x100px
const uint16_t Internet_image [] PROGMEM = {
	0xffff, 0xffff, 0xffff, 0xffff, 0xffff, 0xffff, 0xffff, 0xffff, 0xffff, 0xffff, 0xffff, 0xffff, 0xffff, 0xffff, 0xffff, 0xffff,

**Reply by jlt206** [reply]
Well... I spent the weekend trying to get this to work and found no matter what I do, I can get images to display but not anything I draw.

So, I must be missing something or the 7789 disply is the issue.

Everything works fine with TFT. but not when converted to a sprite.

**Reply by Bodmer** [reply]
16 bit images cannot be plotted into a 4 bit paletted sprite. You will just see garbage.

You need to use an 8 or 16 bit sprite to render a 16 bit image. However an 8 bit sprite only supports 256 colours so the image quality will be poor.

**Reply by jlt206** [reply]
So, I re-read your notes about the hardware/RAM limits on sprites and I can see that I am also running into the limits with the ESP8266 now with these images. So I will need to change my hardware strategy to achieve the results I actually want which may mean not using Arduino.
